mecanum_drive_controller/src/odometry.cpp

```diff
@@ -0,0 +1,123 @@
+  ///       to interpret and compare behavior curves).
+
+  /// \note The variables meaning:
+  /// angular_transformation_from_center_2_base: Rotation transformation matrix, to transform from center frame to base frame
+  /// linear_transformation_from_center_2_base: offset/linear transformation matrix, to transform from center frame to base frame
+
+  double velocity_in_center_frame_linear_x =
+    0.25 * wheels_radius_ * (wheel_front_left_vel + wheel_back_left_vel + wheel_back_right_vel + wheel_front_right_vel);
+  double velocity_in_center_frame_linear_y =
+    0.25 * wheels_radius_ * (-wheel_front_left_vel + wheel_back_left_vel - wheel_back_right_vel + wheel_front_right_vel);
+  double velocity_in_center_frame_angular_z =
+    0.25 * wheels_radius_ / wheels_k_ * (-wheel_front_left_vel - wheel_back_left_vel + wheel_back_right_vel + wheel_front_right_vel);
+
+  tf2::Quaternion orientation_R_c_b;
+  orientation_R_c_b.setRPY(0.0, 0.0, -base_frame_offset_[2]);
+
+  tf2::Matrix3x3 angular_transformation_from_center_2_base = tf2::Matrix3x3((orientation_R_c_b));
+  tf2::Vector3 velocity_in_center_frame_w_r_t_base_frame_ =
+    angular_transformation_from_center_2_base *
+    tf2::Vector3(velocity_in_center_frame_linear_x, velocity_in_center_frame_linear_y, 0.0);
+  tf2::Vector3 linear_transformation_from_center_2_base =
+    angular_transformation_from_center_2_base *
+    tf2::Vector3(-base_frame_offset_[0], -base_frame_offset_[1], 0.0);
+
+  velocity_in_base_frame_linear_x =
+    velocity_in_center_frame_w_r_t_base_frame_.x() +
+    linear_transformation_from_center_2_base.y() * velocity_in_center_frame_angular_z;
+  velocity_in_base_frame_linear_y =
+    velocity_in_center_frame_w_r_t_base_frame_.y() -
+    linear_transformation_from_center_2_base.x() * velocity_in_center_frame_angular_z;
+  velocity_in_base_frame_angular_z = velocity_in_center_frame_angular_z;
+
+  /// Integration.
+  /// NOTE: the position is expressed in the odometry frame , unlike the twist which is
+  ///       expressed in the body frame.
+  orientation_z_in_base_frame_ += velocity_in_base_frame_angular_z * dt;
+
+  tf2::Quaternion orientation_R_b_odom;
+  orientation_R_b_odom.setRPY(0.0, 0.0, -base_frame_offset_[2]);
+
+  tf2::Matrix3x3 angular_transformation_from_base_2_odom = tf2::Matrix3x3((orientation_R_b_odom));
+  tf2::Vector3 velocity_in_base_frame_w_r_t_odom_frame_ =
+    angular_transformation_from_base_2_odom *
+    tf2::Vector3(velocity_in_base_frame_linear_x, velocity_in_base_frame_linear_y, 0.0);
+
+  position_x_in_base_frame_ += velocity_in_base_frame_w_r_t_odom_frame_.x() * dt;
+  position_y_in_base_frame_ += velocity_in_base_frame_w_r_t_odom_frame_.y() * dt;
+  // leaving the below comments intentionally
+  // fprintf(stderr, " position_x_in_base_frame_ = %f  \n", position_x_in_base_frame_);
+  // fprintf(stderr, " position_y_in_base_frame_ = %f  \n", position_y_in_base_frame_);
```

Comment:
If you leave something, then please consider using ROS logging and not `fprintf`. Especially, you are using `stderr` output from information. This doesn't look professional.

```suggestion
```